crates/pathfinder/src/state/block_hash.rs

```diff
@@ -0,0 +1,310 @@
+        // number of transactions
+        StarkHash::from_u64(num_transactions),
+        // transaction commitment
+        transaction_commitment,
+        // number of events
+        StarkHash::from_u64(num_events),
+        // event commitment
+        event_commitment,
+        // reserved: protocol version
+        StarkHash::ZERO,
+        // reserved: extra data
+        StarkHash::ZERO,
+        // parent block hash
+        block.parent_block_hash.0,
+    ];
+
+    let block_hash = stark_hash_of_array(data.into_iter());
+
+    Ok(StarknetBlockHash(block_hash))
+}
+
+#[derive(Default)]
+struct CommitmentTree {
+    tree: MerkleTree<()>,
+}
+
+impl CommitmentTree {
+    pub fn set(&mut self, index: u64, value: StarkHash) -> Result<()> {
+        let key = index.to_be_bytes();
+        self.tree.set(key.view_bits(), value)
+    }
+
+    pub fn commit(self) -> Result<StarkHash> {
+        self.tree.commit()
+    }
+}
+
+fn calculate_transaction_commitment(transactions: &[Transaction]) -> Result<StarkHash> {
+    let mut tree = CommitmentTree::default();
+
+    transactions
+        .iter()
+        .enumerate()
+        .try_for_each(|(idx, tx)| {
+            let idx: u64 = idx.try_into()?;
+            let final_hash = calculate_transaction_hash_with_signature(tx);
+            tree.set(idx, final_hash)?;
+            Result::<_, Error>::Ok(())
+        })
+        .context("Failed to create transaction commitment tree")?;
+
+    tree.commit()
+}
+
+fn calculate_transaction_hash_with_signature(tx: &Transaction) -> StarkHash {
+    lazy_static::lazy_static!(
+        static ref HASH_OF_EMPTY_LIST: StarkHash = stark_hash_of_array([].into_iter());
+    );
+
+    let signature_hash = match &tx.signature {
+        None => *HASH_OF_EMPTY_LIST,
+        Some(signatures) => stark_hash_of_array(signatures.iter().map(|e| e.0.to_owned())),
+    };
+
+    stark_hash(tx.transaction_hash.0, signature_hash)
+}
+
+fn calculate_event_commitment(transaction_receipts: &[Receipt]) -> Result<StarkHash> {
+    let mut tree = CommitmentTree::default();
+
+    transaction_receipts
+        .iter()
+        .flat_map(|receipt| receipt.events.iter())
+        .enumerate()
+        .try_for_each(|(idx, e)| {
+            let idx: u64 = idx.try_into()?;
+            let event_hash = calculate_event_hash(e);
+            tree.set(idx, event_hash)?;
+            Result::<_, Error>::Ok(())
+        })
+        .context("Failed to create event commitment tree")?;
+
+    tree.commit()
+}
+
+fn calculate_event_hash(event: &Event) -> StarkHash {
+    let keys_hash = stark_hash_of_array(event.keys.iter().map(|key| key.0));
+    let data_hash = stark_hash_of_array(event.data.iter().map(|data| data.0));
+
+    stark_hash_of_array([event.from_address.0, keys_hash, data_hash].into_iter())
+}
+
+fn stark_hash_of_array<T: Iterator<Item = StarkHash>>(elements: T) -> StarkHash {
+    // the hash of an array of length n is defined as h(...h((h(0,a1),a2),...,an),n)
```

Comment:
Could link to this being the same as HashChain over at contract_hash but kept separate because changes might come to either.

Reply:
Rationale for leaving separated seems to be missing. It's all nice to read coming a piece of non-trivial looking duplicated implementation to understand if this was done on purpose or not.

Reply:
I've removed the duplicate implementation.